ceno_zkvm/src/bin/e2e.rs

```diff
@@ -77,22 +84,36 @@ fn main() {
     type Pcs = Basefold<GoldilocksExt2, BasefoldRSParams>;
     type ExampleProgramTableCircuit<E> = ProgramTableCircuit<E>;
 
-    // set up logger
-    let (flame_layer, _guard) = FlameLayer::with_file("./tracing.folded").unwrap();
-    let subscriber = Registry::default()
-        .with(
-            fmt::layer()
-                .compact()
-                .with_thread_ids(false)
-                .with_thread_names(false),
-        )
-        .with(
-            EnvFilter::builder()
-                .with_default_directive(LevelFilter::DEBUG.into())
-                .from_env_lossy(),
-        )
-        .with(flame_layer.with_threads_collapsed(true));
-    tracing::subscriber::set_global_default(subscriber).unwrap();
+    // default filter
+    let default_filter = EnvFilter::builder()
+        .with_default_directive(LevelFilter::DEBUG.into())
+        .from_env_lossy();
+
+    // filter by profiling level;
+    // spans with level i contain the field "profiling_{i}"
+    // this restricts statistics to first (args.profiling) levels
+    let profiling_level = args.profiling.unwrap_or(1);
+    let filter_by_profiling_level = filter_fn(move |metadata| {
+        (1..)
+            .map(|i| format!("profiling_{i}"))
+            .take(profiling_level)
+            .any(|field| metadata.fields().field(&field).is_some())
+    });
+
+    let fmt_layer = fmt::layer()
+        .compact()
+        .with_thread_ids(false)
+        .with_thread_names(false)
+        .without_time();
+
+    Registry::default()
+        .with(ForestLayer::default())
+        .with(fmt_layer)
+        // if some profiling granularity is specified, use the profiling filter,
+        // otherwise use the default
+        .with((args.profiling.is_some()).then_some(filter_by_profiling_level))
```

Comment:
The parens look a bit confusing here?

Btw, instead of `is_some().then_some()` you can probably just use `.map(|_| filter_by_profiling_level)`?

Reply:
Removed parens. `.map()` is terser, but not as readable in this case I'd say.